### PR TITLE
BUG: Prevent crash on repr of recursive array

### LINKED ISSUE #8960: BUG: ndarray.__repr__ has no guard against recursion
```
x = np.array([None, None])
x[0] = x
repr(x)
```

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes #8960, by:

* Forbidding reentrance of `array2string` (within a single thread)
* Not calling `np.maximum` on arrays which might be recursive - there's no point setting up an `IntegerFormatter` if we're not yet sure we're working with integers

### CODE REVIEW

File: numpy/core/arrayprint.py [hunk 1]
```diff
@@ -455,7 +495,7 @@ def array2string(a, max_line_width=None, precision=None,
             lst = format_function(arr[0])
         else:
             lst = style(x)
-    elif reduce(product, a.shape) == 0:
+    elif functools.reduce(product, a.shape) == 0:
```
Comment (ahaldane) [inline numpy/core/arrayprint.py]: What's the point of this? (forgive my ignorance)

File: numpy/core/arrayprint.py [hunk 2]
```diff
@@ -337,6 +354,29 @@ def _array2string(a, max_line_width, precision, suppress_small, separator=' ',
     return lst
 
 
+def _recursive_guard(fillvalue='...'):
```
Comment (ahaldane) [inline numpy/core/arrayprint.py]: I think it would be good to add a comment here that this is to account for (for example) self-containing object arrays.

**Review by ahaldane — COMMENTED** [review]
The lambda part looks good.

The wrapper part seems fine too. It doesn't protect against recursion where 'self' changes each iteration, as in:

```python
>>> np.set_printoptions(formatter={'all': lambda x: str(np.array([x]))})
```

but I don't think we need that.

File: numpy/core/arrayprint.py (hunk 1 above)
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: Not polluting the global namespace. Previously there was also `from functools import reduce`.

As for why that's there - `reduce` is no longer a global in python 3

File: numpy/core/arrayprint.py (hunk 2 above)
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: Good point - I'll fix that up when I find the time (or else, you can if you don't want to wait - this PR is editable)

Comment (ahaldane) [inline numpy/core/arrayprint.py]: All right, I will wait. Otherwise it looks good to merge to me.

File: numpy/core/arrayprint.py [hunk 3]
```diff
@@ -337,6 +354,38 @@ def _array2string(a, max_line_width, precision, suppress_small, separator=' ',
...
+            finally:
+                repr_running.discard(key)
+
+        return wrapper
+
+    return decorating_function
+
+
+# gracefully handle recursive calls - this comes up when object arrays contain
+# themselves
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: @ahaldane: Done - seemed more appropriate to comment "why" at the call-site

File: numpy/core/arrayprint.py [hunk 4]
```diff
@@ -337,6 +354,38 @@ def _array2string(a, max_line_width, precision, suppress_small, separator=' ',
...
 
+def _recursive_guard(fillvalue='...'):
+    """
+    Like the python 3.2 reprlib.recursive_repr, but forwards *args and **kwargs
+
+    Decorates a function such that if it calls itself with the same first
+    argument, it returns `fillvalue` instead of recursing.
+
+    Largely copied from reprlib.recursive_repr
+    """
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py]: Docstring improved too.

**Comment by ahaldane** [comment]
LGTM. I'm going to go ahead and merge. Thanks @eric-wieser !

**Comment by eric-wieser** [comment]
Great! Let's see if this makes #8983 easier

**Pull request merged**